File: scripts/deliver.py
#!/usr/bin/env python3
"""Unified message delivery for context-saver pipelines.

Supports multiple backends: iMessage, Telegram, Slack, Discord, webhook.
Each backend is auto-detected based on available CLI tools and env vars.

Usage:
    # iMessage (default if imsg is available)
    python3 deliver.py --to +17029311279 --text "Hello"

    # Telegram
    python3 deliver.py --backend telegram --to 5328771204 --text "Hello"

    # Slack webhook
    python3 deliver.py --backend slack --text "Hello"

    # Auto-detect: tries imsg first, then telegram, then slack
    python3 deliver.py --to +17029311279 --text "Hello"

    # Pipe from stdin (for pipeline chaining)
    echo "Hello" | python3 deliver.py --to +17029311279

Environment variables:
    TELEGRAM_BOT_TOKEN   — Telegram bot token for telegram backend
    TELEGRAM_CHAT_ID     — Default Telegram chat ID
    SLACK_WEBHOOK_URL    — Slack incoming webhook URL
    DISCORD_WEBHOOK_URL  — Discord webhook URL
"""

import argparse
import json
import os
import shutil
import subprocess
import sys
import urllib.request
import urllib.error

OPENCLAW_HOME = os.environ.get("OPENCLAW_HOME", os.path.expanduser("~/.openclaw"))
ENV_FILE = os.path.join(OPENCLAW_HOME, ".env")


def load_env():
    """Load environment variables from .env file."""
    env = os.environ.copy()
    if os.path.exists(ENV_FILE):
        with open(ENV_FILE) as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    key, _, value = line.partition("=")
                    env[key.strip()] = value.strip().strip("\"'")
    return env


# ─────────────────────────────────────────────
# Backend implementations
# ─────────────────────────────────────────────

def send_imessage(to, text, env):
    """Send via imsg CLI."""
    imsg = shutil.which("imsg") or "/opt/homebrew/bin/imsg"
    if not os.path.isfile(imsg):
        return False, "imsg CLI not found"

    # Validate phone number format
    if not to or not to.startswith("+"):
        return False, f"Invalid phone number format: {to} (must start with +)"

    try:
        result = subprocess.run(
            [imsg, "send", "--to", to, "--text", text],
            capture_output=True,
            text=True,
            timeout=15,
        )
        if result.returncode == 0:
            return True, "delivered"
        return False, result.stderr.strip() or f"exit code {result.returncode}"
    except subprocess.TimeoutExpired:
        return False, "imsg send timed out"
    except Exception as e:
        return False, str(e)


def send_telegram(to, text, env):
    """Send via Telegram Bot API (stdlib only, no requests)."""
    token = env.get("TELEGRAM_BOT_TOKEN", "")
    chat_id = to or env.get("TELEGRAM_CHAT_ID", "")

    if not token:
        return False, "TELEGRAM_BOT_TOKEN not set"
    if not chat_id:
        return False, "No chat_id provided and TELEGRAM_CHAT_ID not set"

    url = f"https://api.telegram.org/bot{token}/sendMessage"
    payload = json.dumps({
        "chat_id": chat_id,
        "text": text,
        "parse_mode": "Markdown",
    }).encode("utf-8")

    req = urllib.request.Request(
        url,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            result = json.loads(resp.read())
            if result.get("ok"):
                return True, "delivered"
            return False, result.get("description", "unknown error")
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")[:200]
        return False, f"HTTP {e.code}: {body}"
    except Exception as e:
        return False, str(e)


def send_slack(to, text, env):
    """Send via Slack incoming webhook."""
    webhook_url = env.get("SLACK_WEBHOOK_URL", "")
    if not webhook_url:
        return False, "SLACK_WEBHOOK_URL not set"

    payload = json.dumps({"text": text}).encode("utf-8")
    req = urllib.request.Request(
        webhook_url,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return True, "delivered"
    except urllib.error.HTTPError as e:
        return False, f"HTTP {e.code}"
    except Exception as e:
        return False, str(e)


def send_discord(to, text, env):
    """Send via Discord webhook."""
    webhook_url = env.get("DISCORD_WEBHOOK_URL", "")
    if not webhook_url:
        return False, "DISCORD_WEBHOOK_URL not set"

    payload = json.dumps({"content": text}).encode("utf-8")
    req = urllib.request.Request(
        webhook_url,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return True, "delivered"
    except urllib.error.HTTPError as e:
        return False, f"HTTP {e.code}"
    except Exception as e:
        return False, str(e)


BACKENDS = {
    "imessage": send_imessage,
    "telegram": send_telegram,
    "slack": send_slack,
    "discord": send_discord,
}


def detect_backend(env):
    """Auto-detect the best available backend."""
    # iMessage first (macOS only)
    imsg = shutil.which("imsg") or "/opt/homebrew/bin/imsg"
    if os.path.isfile(imsg):
        return "imessage"

    # Telegram if token exists
    if env.get("TELEGRAM_BOT_TOKEN"):
        return "telegram"

    # Slack if webhook exists
    if env.get("SLACK_WEBHOOK_URL"):
        return "slack"

    # Discord if webhook exists
    if env.get("DISCORD_WEBHOOK_URL"):
        return "discord"

    return None


def deliver(backend, to, text, env):
    """Send a message through the specified backend."""
    if backend not in BACKENDS:
        return False, f"Unknown backend: {backend}. Available: {', '.join(BACKENDS.keys())}"
    return BACKENDS[backend](to, text, env)


def main():
    parser = argparse.ArgumentParser(
        description="Unified message delivery for context-saver pipelines",
        epilog="Backends: imessage, telegram, slack, discord",
    )
    parser.add_argument("--backend", "-b", default="auto",
                        help="Delivery backend (default: auto-detect)")
    parser.add_argument("--to", "-t", action="append",
                        help="Recipient (phone for imessage, chat_id for telegram)")
    parser.add_argument("--text", help="Message text (or pipe via stdin)")
    parser.add_argument("--json", action="store_true",
                        help="Output JSON result")
    args = parser.parse_args()

    env = load_env()

    # Get text from args or stdin
    text = args.text
    if not text and not sys.stdin.isatty():
        text = sys.stdin.read().strip()
    if not text:
        parser.error("--text required (or pipe via stdin)")

    # Detect backend
    backend = args.backend
    if backend == "auto":
        backend = detect_backend(env)
        if not backend:
            print("❌ No delivery backend available", file=sys.stderr)
            print("   Configure one of: imsg CLI, TELEGRAM_BOT_TOKEN, SLACK_WEBHOOK_URL, DISCORD_WEBHOOK_URL",
                  file=sys.stderr)
            sys.exit(1)

    # Send to each recipient
    recipients = args.to or [None]  # None = use default for backend
    results = []
    for recipient in recipients:
        ok, detail = deliver(backend, recipient, text, env)
        results.append({
            "backend": backend,
            "to": recipient,
            "status": "delivered" if ok else "failed",
            "detail": detail,
        })
        if not args.json:
            icon = "✅" if ok else "❌"
            target = recipient or "default"
            print(f"{icon} [{backend}] {target}: {detail}")

    if args.json:
        print(json.dumps(results, indent=2))

    # Exit with error if any delivery failed
    if not all(r["status"] == "delivered" for r in results):
        sys.exit(1)


if __name__ == "__main__":
    main()

File: src/tools/deliver.ts
import { z } from "zod";
import { execSync } from "child_process";
import * as https from "https";
import * as http from "http";
import { loadEnv } from "../lib/env";

export const deliverSchema = z.object({
  text: z.string().describe("Message text to deliver"),
  to: z
    .array(z.string())
    .optional()
    .describe("Recipients (phone numbers, chat IDs). Repeatable."),
  backend: z
    .enum(["auto", "imessage", "telegram", "slack", "discord"])
    .default("auto")
    .describe("Delivery backend (default: auto-detect)"),
});

export type DeliverInput = z.infer<typeof deliverSchema>;

const E164_REGEX = /^\+[1-9]\d{7,14}$/;

function isImsgAvailable(): boolean {
  try {
    execSync("which imsg", { stdio: "ignore" });
    return true;
  } catch {
    return false;
  }
}

function detectBackend(): string {
  const env = loadEnv();
  if (isImsgAvailable()) return "imessage";
  if (env.TELEGRAM_BOT_TOKEN) return "telegram";
  if (env.SLACK_WEBHOOK_URL) return "slack";
  if (env.DISCORD_WEBHOOK_URL) return "discord";
  return "none";
}

function sendImessage(to: string, text: string): { status: string; detail: string } {
  if (!E164_REGEX.test(to)) {
    return { status: "failed", detail: `Invalid phone number: ${to}` };
  }

  try {
    execSync(`imsg send --to "${to}" --text "${text.replace(/"/g, '\\"')}"`, {
      timeout: 15000,
      stdio: "ignore",
    });
    return { status: "delivered", detail: "delivered" };
  } catch (err) {
    return {
      status: "failed",
      detail: err instanceof Error ? err.message : String(err),
    };
  }
}

function postJson(
  url: string,
  payload: Record<string, unknown>
): Promise<{ status: string; detail: string }> {
  return new Promise((resolve) => {
    const data = JSON.stringify(payload);
    const client = url.startsWith("https") ? https : http;
    const urlObj = new URL(url);

    const req = client.request(
      {
        hostname: urlObj.hostname,
        port: urlObj.port,
        path: urlObj.pathname + urlObj.search,
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Content-Length": Buffer.byteLength(data),
        },
        timeout: 10000,
      },
      (res) => {
        let body = "";
        res.on("data", (chunk: Buffer) => {
          body += chunk;
        });
        res.on("end", () => {
          if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
            resolve({ status: "delivered", detail: `HTTP ${res.statusCode}` });
          } else {
            resolve({
              status: "failed",
              detail: `HTTP ${res.statusCode}: ${body.slice(0, 200)}`,
            });
          }
        });
      }
    );
    req.on("error", (err) => {
      resolve({ status: "failed", detail: err.message });
    });
    req.on("timeout", () => {
      req.destroy();
      resolve({ status: "failed", detail: "Request timed out" });
    });
    req.write(data);
    req.end();
  });
}

async function sendTelegram(
  to: string,
  text: string
): Promise<{ status: string; detail: string }> {
  const env = loadEnv();
  const token = env.TELEGRAM_BOT_TOKEN;
  if (!token) return { status: "failed", detail: "TELEGRAM_BOT_TOKEN not set" };

  const chatId = to || env.TELEGRAM_CHAT_ID;
  if (!chatId) return { status: "failed", detail: "No chat ID provided" };

  return postJson(`https://api.telegram.org/bot${token}/sendMessage`, {
    chat_id: chatId,
    text,
    parse_mode: "Markdown",
  });
}

async function sendSlack(
  text: string
): Promise<{ status: string; detail: string }> {
  const env = loadEnv();
  const url = env.SLACK_WEBHOOK_URL;
  if (!url) return { status: "failed", detail: "SLACK_WEBHOOK_URL not set" };

  return postJson(url, { text });
}

async function sendDiscord(
  text: string
): Promise<{ status: string; detail: string }> {
  const env = loadEnv();
  const url = env.DISCORD_WEBHOOK_URL;
  if (!url) return { status: "failed", detail: "DISCORD_WEBHOOK_URL not set" };

  return postJson(url, { content: text });
}

export async function handleDeliver(args: DeliverInput) {
  const backend =
    args.backend === "auto" ? detectBackend() : args.backend;

  if (backend === "none") {
    return {
      content: [
        {
          type: "text" as const,
          text: JSON.stringify({
            success: false,
            error:
              "No delivery backend available. Install imsg (macOS) or set TELEGRAM_BOT_TOKEN, SLACK_WEBHOOK_URL, or DISCORD_WEBHOOK_URL.",
          }),
        },
      ],
    };
  }

  const recipients = args.to || ["default"];
  const results: Array<{
    backend: string;
    to: string;
    status: string;
    detail: string;
  }> = [];

  for (const recipient of recipients) {
    let result: { status: string; detail: string };

    switch (backend) {
      case "imessage":
        result = sendImessage(recipient, args.text);
        break;
      case "telegram":
        result = await sendTelegram(recipient, args.text);
        break;
      case "slack":
        result = await sendSlack(args.text);
        break;
      case "discord":
        result = await sendDiscord(args.text);
        break;
      default:
        result = { status: "failed", detail: `Unknown backend: ${backend}` };
    }

    results.push({ backend, to: recipient, ...result });
  }

  const allDelivered = results.every((r) => r.status === "delivered");

  return {
    content: [
      {
        type: "text" as const,
        text: JSON.stringify({
          success: allDelivered,
          deliveries: results,
        }),
      },
    ],
  };
}
